src/Lecture2.hs

```diff
@@ -74,7 +80,16 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt:: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt 0 (x:xs) = (Just x, xs)
+removeAt n xs
+  | n < 0 = (Nothing, xs)
+  | otherwise = bisect n [] xs
+    where
+      bisect _ _ [] = (Nothing, xs)
+      bisect 0 left (y:ys) = (Just y, left <> ys)
+      bisect position left (y:ys) = bisect (position - 1) (left <> [y]) ys
```

Comment:
It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible. Adding elements to the end of the list is rather slow 🐌